packages/flame_studio/lib/src/widgets/panels/hierarchy_view.dart

```diff
@@ -0,0 +1,239 @@
+import 'package:flame/components.dart';
+import 'package:flame_studio/src/core/component_tree.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+class HierarchyView extends ConsumerStatefulWidget {
+  const HierarchyView({super.key});
+
+  @override
+  HierarchyViewState createState() => HierarchyViewState();
+}
+
+class HierarchyViewState extends ConsumerState<ConsumerStatefulWidget> {
+  Component? selectedComponent;
+  Set<Component> expandedComponents = {};
+
+  @override
+  Widget build(BuildContext context) {
+    final componentTree = ref.watch(componentTreeProvider).root;
+    if (componentTree == null) {
+      return Container();
+    }
+
+    final listItems = <Widget>[];
+    _buildList(componentTree, 0, listItems);
+
+    return ListView(
+      padding: const EdgeInsets.all(8),
+      children: listItems,
+    );
+  }
+
+  void _buildList(ComponentTreeNode node, int depth, List<Widget> out) {
+    out.add(_ListItem(this, node, depth, out.isEmpty));
+    final isExpanded = expandedComponents.contains(node.component);
+    if (isExpanded && node.hasChildren) {
+      for (final childNode in node.children!) {
+        _buildList(childNode, depth + 1, out);
+      }
+      out.add(_ClosingBrace(depth));
+    }
+  }
+
+  /// Used by the expander icons in the gutter: the component expands/collapses
+  /// without affecting the selection state.
+  void toggle(Component component) {
+    setState(() {
+      if (expandedComponents.contains(component)) {
+        expandedComponents.remove(component);
+      } else {
+        expandedComponents.add(component);
+      }
+    });
+  }
+
+  void toggleOrSelect(Component component) {
+    final isExpanded = expandedComponents.contains(component);
+    final isSelected = selectedComponent == component;
+    setState(() {
+      if (isExpanded && isSelected) {
+        expandedComponents.remove(component);
+      } else {
+        expandedComponents.add(component);
+      }
+      selectedComponent = component;
+    });
+  }
+}
+
+class _ListItem extends StatelessWidget {
+  const _ListItem(this.state, this.node, this.indent, this.isFirst);
```

Comment:
Since there will be quite a lot of these ones in a list they should have a `Key` so that it doesn't have to rebuild the whole list all the time. If there is something that a key can be created from?

Reply:
Added a key